File: android/app/src/main/java/com/voixpe3per/pairing/PairingPayload.kt
package com.voixpe3per.pairing

import android.net.Uri
import org.json.JSONObject

data class PairingPayload(
    val mode: String,
    val host: String,
    val port: Int,
    val token: String,
    val relay: String,
    val room: String
) {
    val url: String
        get() = if (mode == "relay") relay else "ws://$host:$port/ws"

    companion object {
        fun fromQr(raw: String): PairingPayload {
            if (raw.startsWith("http://") || raw.startsWith("https://")) {
                return fromUrl(raw)
            }

            val json = JSONObject(raw)
            return PairingPayload(
                mode = json.optString("mode", "lan"),
                host = json.optString("host", ""),
                port = json.optInt("port", 8080),
                token = json.optString("token", ""),
                relay = json.optString("relay", ""),
                room = json.optString("room", "")
            )
        }

        private fun fromUrl(raw: String): PairingPayload {
            val uri = Uri.parse(raw)
            return PairingPayload(
                mode = uri.getQueryParameter("mode") ?: "relay",
                host = uri.getQueryParameter("host") ?: "",
                port = uri.getQueryParameter("port")?.toIntOrNull() ?: 8080,
                token = uri.getQueryParameter("token") ?: "",
                relay = uri.getQueryParameter("relay") ?: "",
                room = uri.getQueryParameter("room") ?: ""
            )
        }
    }
}

File: ios/voiXPe3per/Pairing/PairingPayload.swift
import Foundation

struct PairingPayload: Equatable {
    let mode: String
    let host: String
    let port: Int
    let token: String
    let relay: String
    let publicURL: String
    let room: String

    func webSocketURL() throws -> URL {
        if mode == "relay" || mode == "direct" {
            guard let relayURL = URL(string: relay.isEmpty ? publicURL : relay) else {
                throw PairingError.invalidWebSocketURL
            }
            return relayURL
        }

        guard let lanURL = URL(string: "ws://\(host):\(port)/ws") else {
            throw PairingError.invalidWebSocketURL
        }
        return lanURL
    }

    static func parse(_ raw: String) throws -> PairingPayload {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.hasPrefix("http://") || trimmed.hasPrefix("https://") || trimmed.hasPrefix("voixpe3per://") {
            return try parseURL(trimmed)
        }

        guard let data = trimmed.data(using: .utf8) else {
            throw PairingError.invalidPayload
        }

        let object = try JSONSerialization.jsonObject(with: data)
        guard let json = object as? [String: Any] else {
            throw PairingError.invalidPayload
        }

        return PairingPayload(
            mode: json.string("mode") ?? "lan",
            host: json.string("host") ?? "",
            port: json.int("port") ?? 8080,
            token: json.string("token") ?? "",
            relay: json.string("relay") ?? "",
            publicURL: json.string("public") ?? "",
            room: json.string("room") ?? ""
        )
    }

    private static func parseURL(_ raw: String) throws -> PairingPayload {
        guard let components = URLComponents(string: raw) else {
            throw PairingError.invalidPayload
        }

        return PairingPayload(
            mode: components.queryValue("mode") ?? "relay",
            host: components.queryValue("host") ?? "",
            port: Int(components.queryValue("port") ?? "") ?? 8080,
            token: components.queryValue("token") ?? "",
            relay: components.queryValue("relay") ?? "",
            publicURL: components.queryValue("public") ?? "",
            room: components.queryValue("room") ?? ""
        )
    }
}

enum PairingError: LocalizedError {
    case invalidPayload
    case invalidWebSocketURL
    case serverRejected(String)

    var errorDescription: String? {
        switch self {
        case .invalidPayload:
            return "QR pairing tidak valid"
        case .invalidWebSocketURL:
            return "WebSocket URL tidak valid"
        case .serverRejected(let message):
            return message
        }
    }
}

private extension URLComponents {
    func queryValue(_ name: String) -> String? {
        queryItems?.first { $0.name == name }?.value
    }
}

private extension Dictionary where Key == String, Value == Any {
    func string(_ key: String) -> String? {
        self[key] as? String
    }

    func int(_ key: String) -> Int? {
        if let value = self[key] as? Int {
            return value
        }
        if let value = self[key] as? String {
            return Int(value)
        }
        return nil
    }
}
